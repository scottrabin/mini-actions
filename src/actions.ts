type OptionalProperty<T, P extends {}> =
    (T extends void
        ? unknown
        : (unknown extends T
            ? { [K in keyof P]?: unknown }
            : { [K in keyof P]: T }));

export type Action<T extends string, P, M> =
    { type: T } &
    OptionalProperty<P, { payload: P }> &
    OptionalProperty<M, { meta: M }>;

export type ActionCreator<T extends string, P, M, Args extends Array<any>> =
    ((...args: Args) => Action<T, P, M>) &
    {
        type: T;
    };

/**
 * Creates an annotated action creator which can be used with the various utility
 * methods to enforce type-safe action handling.
 *
 * @param type unique identifier for the action
 * @param toPayload function converting the action creator arguments into the `payload` property
 * @param toMeta function converting the action creator arguments into the `meta` property
 */
export function createAction<T extends string>(
    type: T,
): ActionCreator<T, void, void, Array<void>>;
export function createAction<T extends string, P, Args extends Array<any>>(
    type: T,
    toPayload: (...args: Args) => P,
): ActionCreator<T, P, void, Args>;
export function createAction<T extends string, P, M, Args extends Array<any>>(
    type: T,
    toPayload: null | ((...args: Args) => P),
    toMeta: (...args: Args) => M,
): ActionCreator<T, P, M, Args>;
export function createAction<T extends string, P, M, Args extends Array<any>>(
    type: T,
    toPayload: null | ((...args: Args) => P) = null,
    toMeta: null | ((...args: Args) => M) = null,
): ActionCreator<T, P, M, Args> {
    let fn: (...args: Args) => any;
    if (toPayload && toMeta) {
        fn = (...args) => ({ type, payload: toPayload(...args), meta: toMeta(...args) });
    } else if (toPayload) {
        fn = (...args) => ({ type, payload: toPayload(...args) });
    } else if (toMeta) {
        fn = (...args) => ({ type, meta: toMeta(...args) });
    } else {
        fn = () => ({ type });
    }
    return Object.assign(fn, { type });
}
