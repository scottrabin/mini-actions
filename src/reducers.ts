import {
    Action,
    ActionCreator,
} from "./actions";

/**
 * A reducer of some state S over the given action type(s) in A, which typically
 * becomes a discriminated union after building up well-typed reducers.
 */
export type Reducer<S, A> = (state: S, action: A) => S;

/**
 * Extracts the defined action from a given reducer
 */
type ReducerAction<R extends Reducer<any, any>> = Parameters<R>[1];

/**
 * An enhanced reducer function augmented with a `when` method, which allows
 * for immutably constructing reducers which accept only specific action types.
 */
export type ReducerCreator<S, A> =
    Reducer<void | S, A> & {
        when: <T extends string, P, M>(
            creator: ActionCreator<T, P, M, any>,
            reducer: Reducer<S, Action<T, P, M>>,
        ) => ReducerCreator<S, A | Action<T, P, M>>;
    };

// eslint-disable-next-line no-undef
function mappedReducer<S, ActionMap extends { [key: string]: Reducer<S, Action<typeof key, any, any>> }>(
    initialState: S,
    actionMap: ActionMap,
): ReducerCreator<S, ReducerAction<ActionMap[keyof ActionMap]>> {
    return Object.assign(
        (state: S = initialState, action: ReducerAction<ActionMap[keyof ActionMap]>) => {
            if (action.type in actionMap) {
                return actionMap[action.type](state, action);
            } else {
                return state;
            }
        },
        {
            when: <T extends string, P, M>(creator: ActionCreator<T, P, M, any>, reducer: Reducer<S, Action<T, P, M>>) => {
                return mappedReducer(initialState, { ...actionMap, [creator.type]: reducer });
            }
        }
    );
}

/**
 * Creates a reducer seeded with the given initial state, which has no defined
 * accepted actions. Serves as the starting point for defining well-typed reducers.
 *
 * @param initialState
 */
export function createReducer<S>(initialState: S): ReducerCreator<S, never> {
    return mappedReducer(initialState, {});
}

/**
 * Combines a set of reducers into one parent object reducer which delegates the
 * modifications to each property of the object to the provided reducer.
 *
 * @param reducers Map of object property to reducer governing its value
 */
export function combineReducers<ReducerMap extends { [key: string]: Reducer<any, any> }>(
    reducers: ReducerMap,
): Reducer<void | { [K in keyof ReducerMap]: ReturnType<ReducerMap[K]> }, ReducerAction<ReducerMap[keyof ReducerMap]>> {
    type State = { [K in keyof ReducerMap]: ReturnType<ReducerMap[K]> };
    return (state, action) => {
        const result: Partial<State> = {};
        let changed = false;

        for (let k of Object.keys(reducers)) {
            result[k] = reducers[k](state && state[k], action);
            if (!changed) {
                changed = (state && state[k]) !== result[k];
            }
        }

        return (changed ? result as State : state);
    };
}
