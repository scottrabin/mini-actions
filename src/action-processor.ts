import { Action, ActionCreator } from "./actions";

/**
 * Extensible function which handles specific action types, as well as optionally
 * handling unrecognized actions.
 */
type ActionProcessor<F, O> =
    (F extends Function
        ? (O extends Function ? (F & O) : F)
        : (O extends Function ? O : never)) &
    {
        /**
         * Creates a new action handler which can handle actions of the provided
         * type.
         *
         * @param creator Action creator whose returned action should be handled
         * @param fn handler of the returned action
         */
        when: <T extends string, P, M, RV>(
            creator: ActionCreator<T, P, M, any>,
            fn: (action: Action<T, P, M>) => RV,
        ) => ActionProcessor<F & typeof fn, O>; // eslint-disable-line no-undef
        /**
         * Creates a new action handler which handles any action using the provided function.
         *
         * @param fn Catch-all action handler called when unspecific actions are provided
         */
        otherwise: <RV>(
            fn: (action: Action<string, unknown, unknown>) => RV,
        ) => ActionProcessor<F, typeof fn>; // eslint-disable-line no-undef
    };

function createProcessor<ActionMap extends { [key: string]: (...args: any) => any }, O extends (...args: any) => any>(
    actionMap: ActionMap,
    otherwise?: O,
): ActionProcessor<ActionMap[keyof ActionMap], O> {
    let fn;
    if (otherwise) {
        fn = <T extends string>(
            action: (T extends keyof ActionMap ? Parameters<ActionMap[T]>[0] : Action<string, unknown, unknown>),
        ): (T extends keyof ActionMap ? ReturnType<ActionMap[T]> : ReturnType<O>) => {
            if (action.type in actionMap) {
                return actionMap[action.type](action);
            } else {
                return otherwise(action);
            }
        };
    } else {
        fn = <T extends keyof ActionMap>(action: Parameters<ActionMap[T]>[0]): ReturnType<ActionMap[T]> => {
            return actionMap[action.type](action);
        };
    }
    return Object.assign(fn, {
        when: <T extends string, P, M, RV>(
            creator: ActionCreator<T, P, M, any>,
            fn: (action: Action<T, P, M>) => RV,
        ) => {
            return createProcessor({ ...actionMap, [creator.type]: fn }, otherwise);
        },
        otherwise: <RV>(
            fn: (action: Action<string, unknown, unknown>) => RV,
        ): ActionProcessor<ActionMap[keyof ActionMap], typeof fn> => {
            return createProcessor(actionMap, fn);
        }
    });
}

/**
 * Creates an action handler which can process actions created by the provided
 * action creator.
 *
 * @param creator Action creator whose returned action should be handled
 * @param fn handler of the returned action
 */
export function when<T extends string, P, M, RV>(
    creator: ActionCreator<T, P, M, any>,
    fn: (action: Action<T, P, M>) => RV,
): ActionProcessor<(action: Action<T, P, M>) => RV, void> {
    return createProcessor({ [creator.type]: fn });
}

/**
 * Creates an action handler which handles any action using the provided function.
 *
 * @param fn Catch-all action handler called when unspecific actions are provided
 */
export function otherwise<RV>(
    fn: (action: Action<string, unknown, unknown>) => RV,
): ActionProcessor<void, typeof fn> {
    return createProcessor({}, fn);
}
