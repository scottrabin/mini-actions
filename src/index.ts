import { when, otherwise } from "./action-processor";

export {
    Action,
    createAction,
} from "./actions";
export const handle = {
    when,
    otherwise,
};
export {
    Reducer,
    createReducer,
    combineReducers,
} from "./reducers";
