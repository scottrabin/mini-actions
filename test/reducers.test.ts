import { createAction } from "../src/actions";
import { createReducer, combineReducers } from "../src/reducers";

describe("reducers", () => {
    describe("createReducer", () => {
        it("creates a reducer which returns the seeded initial state when given no current state", () => {
            const initialState = { one: 1 };
            const action = createAction("do nothing");
            const reducer = createReducer(initialState).when(action, s => s);

            expect(reducer(void 0, action())).toBe(initialState);
        });

        it("allows chaining with provided action creators to define a reducer", () => {
            const initialState = { value: 0 };
            const increment = createAction("increment", (value: number) => ({ value }));
            const decrement = createAction("decrement", (value: number) => ({ value }));
            const reducer = createReducer(initialState).
                when(increment, (state, { payload }) => ({
                    value: state.value + payload.value,
                })).
                when(decrement, (state, { payload }) => ({
                    value: state.value - payload.value,
                }));

            expect(reducer(void 0, increment(0))).toEqual({ value: 0 });
            expect(reducer({ value: 1 }, increment(3))).toEqual({ value: 4 });
            expect(reducer({ value: 1 }, decrement(7))).toEqual({ value: -6 });
        });
    });

    describe("combineReducers", () => {
        it("can combine reducers accepting the same action", () => {
            const action = createAction("reset");
            const rOne = createReducer({ one: 1 }).when(action, s => s);
            const rTwo = createReducer({ two: 2 }).when(action, s => s);
            const reducer = combineReducers({ rOne, rTwo });

            expect(reducer(void 0, action())).toStrictEqual({
                rOne: {
                    one: 1,
                },
                rTwo: {
                    two: 2,
                },
            });
            expect(reducer(void 0, action())).toEqual(reducer(void 0, action()));
        });

        it("can combine reducers accepting different actions", () => {
            const moveX = createAction("moveX", (value: number) => ({ value }));
            const moveY = createAction("moveY", (value: number) => ({ value }));
            const xPosition = createReducer(0).
                when(moveX, (state, { payload }) => (state + payload.value));
            const yPosition = createReducer(0).
                when(moveY, (state, { payload }) => (state + payload.value));
            const position = combineReducers({ x: xPosition, y: yPosition });

            expect(position(void 0, moveX(3))).toStrictEqual({ x: 3, y: 0 });
            expect(position(void 0, moveY(-6))).toStrictEqual({ x: 0, y: -6 });
            expect([
                moveX(3),
                moveY(4),
                moveX(-5),
                moveY(-11),
            ].reduce(position, void 0)).toStrictEqual({ x: -2, y: -7 });
        });

        it("should return the same initial object if none of the properties have changed", () => {
            const increment = createAction("increment", (value: number) => value);
            const reducer = combineReducers({
                x: createReducer(0).when(increment, (state, { payload }) => state + payload),
            });
            const state = { x: 5 };
            expect(reducer(state, increment(0))).toBe(state);
        });

        it("should short circuit if none of the reducers handle the provided action", () => {
            const firstAction = createAction("first");
            const secondAction = createAction("second");
            const spies = [
                jest.fn(state => state + 1),
                jest.fn(state => state - 1),
            ];
            const reducer = combineReducers({
                top: combineReducers({
                    one: createReducer(0).when(firstAction, spies[0]),
                }),
                bottom: combineReducers({
                    second: createReducer(0).when(secondAction, spies[1]),
                }),
            });

            expect(reducer({ top: { one: 0 }, bottom: { second: 0 } }, secondAction())).toMatchSnapshot();
            expect(spies[0].mock.calls.length).toBe(0);
            expect(spies[1].mock.calls.length).toBe(1);
        });

        it("should accept reducers not created via `createReducer`", () => {
            const action = createAction("action");
            const reducer = combineReducers({
                // eslint-disable-next-line @typescript-eslint/no-unused-vars
                property: (state: number = 0, _: any) => state,
            });

            expect(reducer(void 0, action())).toMatchSnapshot();
        });
    });
});
