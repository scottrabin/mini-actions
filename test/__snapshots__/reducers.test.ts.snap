// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers combineReducers should accept reducers not created via \`createReducer\` 1`] = `
Object {
  "property": 0,
}
`;

exports[`reducers combineReducers should short circuit if none of the reducers handle the provided action 1`] = `
Object {
  "bottom": Object {
    "second": -1,
  },
  "top": Object {
    "one": 0,
  },
}
`;
